src/analysis/processing/qgsmetadataalgorithms.cpp

```diff
@@ -248,4 +248,311 @@ QVariantMap QgsExportLayerMetadataAlgorithm::processAlgorithm( const QVariantMap
+{
+  return QObject::tr( "Metadata tools" );
+}
+
+QString QgsAddHistoryMetadataAlgorithm::groupId() const
+{
+  return QStringLiteral( "metadatatools" );
+}
+
+QString QgsAddHistoryMetadataAlgorithm::shortHelpString() const
+{
+  return QObject::tr( "Adds a new history entry to the layer's metadata." );
+}
+
+QgsAddHistoryMetadataAlgorithm *QgsAddHistoryMetadataAlgorithm::createInstance() const
+{
+  return new QgsAddHistoryMetadataAlgorithm();
+}
+
+void QgsAddHistoryMetadataAlgorithm::initAlgorithm( const QVariantMap & )
+{
+  addParameter( new QgsProcessingParameterMapLayer( QStringLiteral( "INPUT" ), QObject::tr( "Layer" ) ) );
+  addParameter( new QgsProcessingParameterString( QStringLiteral( "HISTORY" ), QObject::tr( "History entry" ) ) );
+  addOutput( new QgsProcessingOutputMapLayer( QStringLiteral( "OUTPUT" ), QObject::tr( "Updated" ) ) );
+}
+
+bool QgsAddHistoryMetadataAlgorithm::prepareAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback * )
+{
+  QgsMapLayer *layer = parameterAsLayer( parameters, QStringLiteral( "INPUT" ), context );
+  const QString history = parameterAsString( parameters, QStringLiteral( "HISTORY" ), context );
+
+  if ( !layer )
+    throw QgsProcessingException( QObject::tr( "Invalid input layer" ) );
+
+  mLayerId = layer->id();
+
+  std::unique_ptr<QgsLayerMetadata> md( layer->metadata().clone() );
+  md->addHistoryItem( history );
+  layer->setMetadata( *md.get() );
+
+  return true;
+}
+
+QVariantMap QgsAddHistoryMetadataAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback * )
+{
+  Q_UNUSED( parameters );
+  Q_UNUSED( context );
+
+  QVariantMap results;
+  results.insert( QStringLiteral( "OUTPUT" ), mLayerId );
+  return results;
+}
+
+///
+
+QString QgsUpdateLayerMetadataAlgorithm::name() const
+{
+  return QStringLiteral( "updatelayermetadata" );
+}
+
+QString QgsUpdateLayerMetadataAlgorithm::displayName() const
+{
+  return QObject::tr( "Update layer metadata" );
+}
+
+QStringList QgsUpdateLayerMetadataAlgorithm::tags() const
+{
+  return QObject::tr( "change,update,layer,metadata,qmd" ).split( ',' );
+}
+
+QString QgsUpdateLayerMetadataAlgorithm::group() const
+{
+  return QObject::tr( "Metadata tools" );
+}
+
+QString QgsUpdateLayerMetadataAlgorithm::groupId() const
+{
+  return QStringLiteral( "metadatatools" );
+}
+
+QString QgsUpdateLayerMetadataAlgorithm::shortHelpString() const
+{
+  return QObject::tr( "Copies all non-empty metadata fields from an input layer to a target layer.\n\nLeaves empty input fields unchaged in the target." );
+}
+
+QgsUpdateLayerMetadataAlgorithm *QgsUpdateLayerMetadataAlgorithm::createInstance() const
+{
+  return new QgsUpdateLayerMetadataAlgorithm();
+}
+
+void QgsUpdateLayerMetadataAlgorithm::initAlgorithm( const QVariantMap & )
+{
+  addParameter( new QgsProcessingParameterMapLayer( QStringLiteral( "INPUT" ), QObject::tr( "Source layer" ) ) );
```

Comment:
I'd probably go with "SOURCE" here, as "INPUT" is a bit ambiguous (either layer could reasonably be assumed to be the input here)

```suggestion
  addParameter( new QgsProcessingParameterMapLayer( QStringLiteral( "SOURCE" ), QObject::tr( "Source layer" ) ) );
```

Reply:
Fixed and used the same approach in Copy Metadata algorithm